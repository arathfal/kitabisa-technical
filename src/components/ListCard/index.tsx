'use client'
import { useMemo } from 'react'

import useFilter from '@/store/filter'
import { Donation } from '@/types/donation'

import Card from '../Card'

type ListCardProps = {
  list?: Array<Donation>
}

export default function ListCard({ list = [] }: ListCardProps) {
  const { sort } = useFilter()

  const sortedList = useMemo(() => {
    const isSorted = sort !== 'default'
    const sortByDaysLeft = list?.sort((a, b) => a.days_remaining - b.days_remaining)
    const newList = Array.from(sortByDaysLeft)

    if (isSorted) {
      return newList?.sort((a, b) => {
        const goalA = a?.donation_target
        const goalB = b?.donation_target
        if (sort === 'asc') {
          return goalA - goalB
        }

        return goalB - goalA
      })
    }

    return newList
  }, [list, sort])

  return (
    <div
      data-testid="list-card-test"
      className="grid grid-cols-1 gap-x-6 gap-y-8 sm:grid-cols-2 lg:grid-cols-3"
    >
      {sortedList?.map((item) => <Card key={item?.title} data={item} />)}
    </div>
  )
}
