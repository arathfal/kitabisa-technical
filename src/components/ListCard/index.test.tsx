import { cleanup, render } from '@testing-library/react'

import ListCard from '.'

const mock = [
  {
    id: 94597,
    order: 1,
    parent_project_id: 0,
    title: '#BisaBangkit Bersama Kitabisa',
    expired: 2147483647,
    image: 'https://img.staging.kitabisa.cc/size/664x357/0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg',
    days_remaining: 0,
    donation_received: 178613497,
    campaigner: 'Kitabisa.com',
    campaigner_type: 'PERSONAL',
    campaigner_badge: 'https://assets.kitabisa.com/images/icon__verified-user.svg',
    campaigner_is_verified: true,
    category_name: 'Bencana Alam',
    is_forever_running: true,
    is_open_goal: false,
    request_userdata: false,
    donation_target: 500000000,
    donation_percentage: 0.357227,
    short_url: 'bisabangkit',
    is_featured: 0,
    custom_fb_pixel: ''
  },
  {
    id: 29755,
    order: 2,
    parent_project_id: 0,
    title: 'ponpes al-muawanah tasikmalaya kebakaran',
    expired: 2147483647,
    image:
      'https://kitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com/_production/campaign/29755/31__1489611820_67821505206_f.jpg',
    days_remaining: 0,
    donation_received: 600711,
    campaigner: 'Saepul Aziz',
    campaigner_type: 'PERSONAL',
    campaigner_badge: '',
    campaigner_is_verified: false,
    category_name: 'Bencana Alam',
    is_forever_running: true,
    is_open_goal: false,
    request_userdata: false,
    donation_target: 300000000,
    donation_percentage: 0.00200237,
    short_url: 'bantuankebakaran',
    is_featured: 0,
    custom_fb_pixel: ''
  }
]

const mockSetSort = jest.fn()

jest.mock('@/store/filter', () => ({
  __esModule: true,
  default: jest.fn(() => ({
    sort: 'asc',
    setSort: mockSetSort
  }))
}))

describe('ListCard Component', () => {
  beforeEach(() => mockSetSort.mockClear())
  afterEach(cleanup)

  test('renders the list-card-test component correctly', () => {
    const { getByTestId, asFragment } = render(<ListCard list={mock} />)

    expect(getByTestId('list-card-test')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })
})
