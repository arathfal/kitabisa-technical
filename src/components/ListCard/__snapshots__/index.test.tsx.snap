// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListCard Component renders the list-card-test component correctly 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-1 gap-x-6 gap-y-8 sm:grid-cols-2 lg:grid-cols-3"
    data-testid="list-card-test"
  >
    <div
      data-testid="card-test"
    >
      <div
        class="relative mb-2 h-64 w-full overflow-hidden"
      >
        <img
          alt="ponpes al-muawanah tasikmalaya kebakaran"
          class="object-cover"
          data-nimg="fill"
          decoding="async"
          fetchpriority="high"
          sizes="100%"
          src="/_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=3840&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=16&q=75 16w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=32&q=75 32w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=48&q=75 48w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=64&q=75 64w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=96&q=75 96w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=128&q=75 128w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fkitabisa-userupload-01.s3-ap-southeast-1.amazonaws.com%2F_production%2Fcampaign%2F29755%2F31__1489611820_67821505206_f.jpg&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
      <h2
        class="mb-2 h-auto text-xl capitalize antialiased sm:mb-0 sm:h-16"
      >
        <strong>
          ponpes al-muawanah tasikmalaya kebakaran
        </strong>
      </h2>
      <div
        class="relative w-full overflow-hidden rounded-full bg-slate-300"
      >
        <div
          class="h-2 text-right bg-gray-500"
          style="width: 0.2%;"
        />
      </div>
      <div
        class="mt-2 flex items-center justify-between gap-2"
      >
        <div
          class="flex flex-col"
        >
          <h6
            class="font-semibold"
          >
            Terkumpul
          </h6>
          <p>
            Rp. 600.711
          </p>
        </div>
        <div
          class="flex flex-col"
        >
          <h6
            class="font-semibold"
          >
            Sisa Hari
          </h6>
          <p
            class="text-right"
          >
            0
          </p>
        </div>
      </div>
    </div>
    <div
      data-testid="card-test"
    >
      <div
        class="relative mb-2 h-64 w-full overflow-hidden"
      >
        <img
          alt="#BisaBangkit Bersama Kitabisa"
          class="object-cover"
          data-nimg="fill"
          decoding="async"
          fetchpriority="high"
          sizes="100%"
          src="/_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=3840&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=16&q=75 16w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=32&q=75 32w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=48&q=75 48w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=64&q=75 64w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=96&q=75 96w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=128&q=75 128w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimg.staging.kitabisa.cc%2Fsize%2F664x357%2F0f9a7205-79ef-49c9-a95a-49347fbd00a6.jpg&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
      <h2
        class="mb-2 h-auto text-xl capitalize antialiased sm:mb-0 sm:h-16"
      >
        <strong>
          #BisaBangkit Bersama Kitabisa
        </strong>
      </h2>
      <div
        class="relative w-full overflow-hidden rounded-full bg-slate-300"
      >
        <div
          class="h-2 text-right bg-gray-500"
          style="width: 35.72%;"
        />
      </div>
      <div
        class="mt-2 flex items-center justify-between gap-2"
      >
        <div
          class="flex flex-col"
        >
          <h6
            class="font-semibold"
          >
            Terkumpul
          </h6>
          <p>
            Rp. 178.613.497
          </p>
        </div>
        <div
          class="flex flex-col"
        >
          <h6
            class="font-semibold"
          >
            Sisa Hari
          </h6>
          <p
            class="text-right"
          >
            0
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
